src/script/sign.h

```diff
@@ -61,6 +64,429 @@ struct SignatureData {
     explicit SignatureData(const CScript& script) : scriptSig(script) {}
 };
 
+
+
+// Note: These constants are in reverse byte order because serialization uses LSB
+static constexpr uint32_t PSBT_MAGIC_BYTES = 0x74627370;
+static constexpr uint8_t PSBT_UNSIGNED_TX_NON_WITNESS_UTXO = 0x00;
+static constexpr uint8_t PSBT_REDEEMSCRIPT_WITNESS_UTXO = 0x01;
+static constexpr uint8_t PSBT_WITNESSSCRIPT_PARTIAL_SIG = 0x02;
+static constexpr uint8_t PSBT_BIP32_KEYPATH_SIGHASH = 0x03;
+static constexpr uint8_t PSBT_NUM_IN_VIN = 0x04;
+
+// The separator is 0x00. Reading this in means that the unserializer can interpret it
+// as a 0 length key. which indicates that this is the separator. The separator has no value.
+static const uint8_t PSBT_SEPARATOR = 0x00;
+
+template<typename... X>
+std::vector<unsigned char> SerializeToVector(const X&... args)
+{
+    std::vector<unsigned char> ret;
+    CVectorWriter ss(SER_NETWORK, PROTOCOL_VERSION, ret, 0);
+    SerializeMany(ss, args...);
+    return ret;
+}
+
+/** A structure for PSBTs which contain per input information */
+struct PartiallySignedInput
+{
+    CTransactionRef non_witness_utxo;
+    CTxOut witness_utxo;
+    std::map<CPubKey, std::vector<unsigned char>> partial_sigs;
+    std::map<std::vector<unsigned char>, std::vector<unsigned char>> unknown;
+    int sighash_type = 0;
+    uint64_t index = 0;
+    bool use_in_index = false;
+
+    PartiallySignedInput() {}
+    void SetNull();
+    bool IsNull();
+};
+
+/** A version of CTransaction with the PSBT format*/
+struct PartiallySignedTransaction
+{
+    CMutableTransaction tx;
+    std::map<uint160, CScript> redeem_scripts;
+    std::map<uint256, CScript> witness_scripts;
+    std::vector<PartiallySignedInput> inputs;
+    std::map<std::vector<unsigned char>, std::vector<unsigned char>> unknown;
+    std::map<CPubKey, std::vector<uint32_t>> hd_keypaths;
+    uint64_t num_ins = 0;
+    bool use_in_index = false;
+
+    void SetNull();
+    bool IsNull();
+    PartiallySignedTransaction() {}
+    PartiallySignedTransaction(const CMutableTransaction& tx, const std::map<uint160, CScript>& redeem_scripts, const std::map<uint256, CScript>& witness_scripts, const std::vector<PartiallySignedInput>& inputs, const std::map<CPubKey, std::vector<uint32_t>> hd_keypaths);
+    PartiallySignedTransaction(const PartiallySignedTransaction& psbt_in);
+
+    // Only checks if they refer to the same transaction
+    friend bool operator==(const PartiallySignedTransaction& a, const PartiallySignedTransaction &b)
+    {
+        // Check the inputs
+        for (unsigned int i = 0; i < a.tx.vin.size(); ++i) {
```

Comment:
First check?
```cpp
if (a.tx.vin.size() != b.tx.vin.size() || a.tx.vout.size() != b.tx.vout.size()) return false;
```
Otherwise this allows `b` to have more in/outputs and if `b` has fewer then it oob.

Reply:
Done